pkg/distributor/forwarding/forwarding.go

```diff
@@ -94,14 +110,69 @@ func NewForwarder(cfg Config, reg prometheus.Registerer, log log.Logger) Forward
 			Help:      "The client-side latency of requests to forward metrics made by the Distributor.",
 			Buckets:   []float64{.005, .01, .025, .05, .1, .25, .5, 1, 2.5, 5, 10, 20, 30},
 		}),
+
+		grpcClientsGauge: promauto.With(reg).NewGauge(prometheus.GaugeOpts{
+			Name: "cortex_distributor_forward_grpc_clients",
+			Help: "Number of gRPC clients used by Distributor forwarder.",
+		}),
 	}
 
+	f.httpGrpcClientPool = f.newHTTPGrpcClientsPool()
 	f.Service = services.NewIdleService(f.start, f.stop)
 
 	return f
 }
 
+func (f *forwarder) newHTTPGrpcClientsPool() *client.Pool {
+	poolConfig := client.PoolConfig{
+		CheckInterval: 5 * time.Second,
+
+		// There may be multiple hosts behind single PoolClient, we don't want to close entire client if one of hosts
+		// is unhealthy.
+		HealthCheckEnabled: false,
+	}
+
+	return client.NewPool("forwarding-httpgrpc", poolConfig, nil, f.createHTTPGrpcClient, f.grpcClientsGauge, f.log)
+}
+
+func (f *forwarder) createHTTPGrpcClient(addr string) (client.PoolClient, error) {
+	opts, err := f.cfg.GRPCClientConfig.DialOption(nil, nil)
+
+	if err != nil {
+		return nil, err
+	}
+
+	const grpcServiceConfig = `{"loadBalancingPolicy":"round_robin"}`
+
+	opts = append(opts,
+		grpc.WithBlock(),
+		grpc.WithDefaultServiceConfig(grpcServiceConfig),
+
+		// These settings are quite low, and require server to be tolerant to them.
+		grpc.WithKeepaliveParams(keepalive.ClientParameters{
+			Time:                time.Second * 10,
+			Timeout:             time.Second * 5,
+			PermitWithoutStream: true,
+		}),
+	)
+
+	conn, err := grpc.Dial(addr, opts...)
+	if err != nil {
+		return nil, err
+	}
+
+	return &grpcHTTPClient{
+		HTTPClient:   httpgrpc.NewHTTPClient(conn),
+		HealthClient: grpc_health_v1.NewHealthClient(conn),
+		conn:         conn,
+	}, nil
+}
+
 func (f *forwarder) start(ctx context.Context) error {
+	if err := services.StartAndAwaitRunning(ctx, f.httpGrpcClientPool); err != nil {
+		return errors.Wrap(err, "failed to start client pool")
```

Comment:
Just to avoid confusion with the http client
```suggestion
		return errors.Wrap(err, "failed to start grpc client pool")
```